tools/shared.py

```diff
@@ -1637,35 +1604,94 @@ def closure_compiler(filename, pretty=True):
   @staticmethod
   def is_ar(filename):
     try:
-      if Building._is_ar_cache.get(filename):
+      if filename in Building._is_ar_cache:
         return Building._is_ar_cache[filename]
-      b = open(filename, 'r').read(8)
-      sigcheck = b[0] == '!' and b[1] == '<' and \
-                 b[2] == 'a' and b[3] == 'r' and \
-                 b[4] == 'c' and b[5] == 'h' and \
-                 b[6] == '>' and ord(b[7]) == 10
+      with open(filename, 'r') as f:
+        b = f.read(8)
+        sigcheck = len(b) == 8 and \
+                   b[0] == '!' and b[1] == '<' and \
+                   b[2] == 'a' and b[3] == 'r' and \
+                   b[4] == 'c' and b[5] == 'h' and \
+                   b[6] == '>' and ord(b[7]) == 10
       Building._is_ar_cache[filename] = sigcheck
       return sigcheck
     except Exception, e:
       logging.debug('Building.is_ar failed to test whether file \'%s\' is a llvm archive file! Failed on exception: %s' % (filename, e))
       return False
 
+  _is_bitcode_cache = {}
   @staticmethod
   def is_bitcode(filename):
-    # look for magic signature
-    b = open(filename, 'r').read(4)
-    if b[0] == 'B' and b[1] == 'C':
-      return True
-    # look for ar signature
-    elif Building.is_ar(filename):
-      return True
-    # on OS X, there is a 20-byte prefix
-    elif ord(b[0]) == 222 and ord(b[1]) == 192 and ord(b[2]) == 23 and ord(b[3]) == 11:
-      b = open(filename, 'r').read(24)
-      return b[20] == 'B' and b[21] == 'C'
+    try:
+      if filename in Building._is_bitcode_cache:
+        return Building._is_bitcode_cache[filename]
+      sigcheck = False
+      # look for magic signature
+      with open(filename, 'r') as f:
+        b = f.read(4)
+        if len(b) < 4: return False
+        if len(b) >= 2 and b[0] == 'B' and b[1] == 'C':
+          sigcheck = True
+        # look for ar signature
+        elif Building.is_ar(filename):
+          sigcheck = True
+        # on OS X, there is a 20-byte prefix
+        elif len(b) == 4 and ord(b[0]) == 222 and ord(b[1]) == 192 and \
+                             ord(b[2]) == 23 and ord(b[3]) == 11:
+          b = f.read(18)
+          sigcheck = len(b) == 18 and b[16] == 'B' and b[17] == 'C'
+      Building._is_bitcode_cache[filename] = sigcheck
+      return sigcheck
+    except Exception, e:
+      logging.debug('Building.is_bitcode failed on file \'%s\'! Failed on exception: %s' % (filename, e))
+      return False
 
+  @staticmethod
+  def is_jslibrary(filename):
+    if not filename.endswith(('.jso','.jso.o')):
```

Comment:
We support both `.jso` and `.jso.o` because CMake has an annoying habit of tacking ".o" onto the end of everything. If you tell it to treat `.jso` files as input files, it works as hard as it can to produce `.jso.o` output, and I couldn't find a way to stop it (it seems fairly baked in behaviour).

Rather than fight the system, I just made emcc shrug and handle `.jso.o`, which makes integrating JS libraries very easy into projects using CMake.